### PR TITLE
#1: Basic error object

### LINKED ISSUE #1: Basic error message builder
## Acceptance Criteria

1. Basic builder supports: description, 0..n mitigations, 0..n parameter descriptions
2. Builder can produce a string

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #1.

### CODE REVIEW

File: doc/changes/changelog.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Changelog
```
Comment (jakobbraun) [inline doc/changes/changelog.md:1]:
```suggestion
# Changelog
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,59 @@
...
+```lua
+local errobj = exaerror.create("E-IO-13", "Need %d MiB space, but only %d MiB left on device %s.",
+    500.2, 14.8, "/dev/sda4")s
+```
+
+Use it as string:
+
+```
+print(errobj)
+```
```
Comment (jakobbraun) [inline README.md:22]:
````suggestion
```lua
print(errobj)
```
````

File: README.md [hunk 3]
```diff
@@ -1,2 +1,59 @@
...
+
+Raise a corresponding Lua error
+
+```
+errobj.raise()
+```
```
Comment (jakobbraun) [inline README.md:28]:
````suggestion
```lua
errobj.raise()
```
````

File: README.md [hunk 4]
```diff
@@ -1,2 +1,59 @@
...
+```lua
+exaerror.error("E-IO-13", "Need %d MiB space, but only %d MiB left on device %s.",
+     500.2, 14.8, "/dev/sda4")
+```
+
+Check out the [user guide](doc/user_guide/user_guide.md) for more details.
+
+## Features
+
+1. Define error objects with error code, description, placeholders, parameters and mitigations
```
Comment (jakobbraun) [inline README.md:41]:
```suggestion
1. Define error objects with error code, message, placeholders, parameters and mitigations
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,59 @@
...
+1. Use error objects where strings are expected
+1. Raise errors from error objects
+
+## Table of Contents
```
Comment (jakobbraun) [inline README.md:45]: I would remove this headline

File: README.md [hunk 6]
```diff
@@ -1,2 +1,59 @@
...
+
+* [User Guide](doc/user_guide/user_guide.md)
+* [Change Log](doc/changes/changelog.md)
+
+You can find corresponding libraries for other languages here:
+
+* [Error reporting Java](https://github.com/exasol/error-reporting-java)
+* [Error reporting C#](https://github.com/exasol/error-reporting-csharp)
+
+#### Dependencies
```
Comment (jakobbraun) [inline README.md:57]: Missing table with dependencies / link to dependencies.md file

File: doc/changes/changes_1.0.0.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# error-reporting-lua 1.0.0, released yyyy-mm-dd
+
+Code name: MVP
```
Comment (jakobbraun) [inline doc/changes/changes_1.0.0.md:3]: don't abreviate

File: doc/user_guide/user_guide.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+And you can conveniently rais a Lua `error` from them.
+
+Additionally the resulting code is made to be parseable, so that you can extract an error catalog from the code.
+
+# Defining Error Objects
+
+The core concept of the module is an error object defined in `exaerror`. This is a Lua table that has a couple of predefined attributes:
+
+* `code`: A machine readable error code
```
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:13]: On the confluence page you called something like `E-TEST-1` error-identifier.
In the other projects we try to use this terminilogy.

File: doc/user_guide/user_guide.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+As you can see in the example above, you can simply pass the error attributes as table when calling `exaerror:new`.
+
+The predefined table keys are
+
+* `code`: machine-readable unique code (string)
+* `message`: Error description either as static string or as [format with placeholders](https://www.lua.org/manual/5.1/manual.html#pdf-string.format) (string)
+* `parameters`: parameter values that will be used to replace placeholders in the description (array of values)
+* `mitigations`: list of hints on how to fix the error (array of strings)
```
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:54]: Is it also possible to use placeholder in mitigations (I think it should)

File: doc/user_guide/user_guide.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+})
+```
+
+As you can see in the example above, you can simply pass the error attributes as table when calling `exaerror:new`.
+
+The predefined table keys are
+
+* `code`: machine-readable unique code (string)
+* `message`: Error description either as static string or as [format with placeholders](https://www.lua.org/manual/5.1/manual.html#pdf-string.format) (string)
+* `parameters`: parameter values that will be used to replace placeholders in the description (array of values)
```
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:53]: How can you define the description for the parameters using this API?

File: doc/user_guide/user_guide.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+Mind the colon!
+
+## Creating an Raising an Error in one Step
+
+The shortest variant to create and raise an error in one call is using `exaerror.error`.
+
+```lua
+exaerror.error("E-IO-13", "Need %d MiB space, but only %d MiB left on device %s.",
```
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:99]: I think this will discourage users from adding mitigations.

File: src/exaerror.lua [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+--
+-- @return string representation of the error object
+--
+function M:__tostring()
+    local pieces = {}
+    if self.message then
+        table.insert(pieces, self:get_message())
+    end
+    if self.mitigations then
+        for _, mitigation in ipairs(self.mitigations) do
```
Comment (jakobbraun) [inline src/exaerror.lua:23]: Since you don't use the index, why can't you just iterate the mitigations? Like:
```lua
for mitigation in self.mitigations do
```

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/exaerror.lua [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            table.insert(pieces, mitigation)
```
Comment (jakobbraun) [inline src/exaerror.lua:24]: In Java for multiple mitigations we print:

```
Mitigations:
* Mitigation a
* Mitigation n
```

File: src/exaerror.lua [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        end
+    end
+    if self.code then
+        if #pieces > 0 then
+            table.insert(pieces, 1, self.code .. ":")
```
Comment (jakobbraun) [inline src/exaerror.lua:29]: I'm not sure if you really need to handle this edge case. If someone forgets the message,  I think it's ok to have not so nice error message.
In java we do:
`E-TEST-1: Unkown Column !!!MISSING PARAMETER!!!`
That has the advantage that it catches your attention while testing

Maybe here you could do `E-TEST-1: No Error Message`

File: src/exaerror.lua [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    end
+end
+
+---
+-- Concatenate an error object with another object.
+--
+-- @return String representing the concatenation.
+--
+function M.__concat(left, right)
+    return tostring(left) .. tostring(right)
```
Comment (jakobbraun) [inline src/exaerror.lua:47]: Why do you need this?

File: src/exaerror.lua (hunk 12 above)
Comment (jakobbraun) [reply src/exaerror.lua:23]: I think you're missing parameter replacement here.

File: target/luaunit-reports/test_errormsg.xml [hunk 16]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (jakobbraun) [inline target/luaunit-reports/test_errormsg.xml:1]: Is it intended to commit these files?

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: README.md [hunk 17]
```diff
@@ -1,2 +1,59 @@
 # error-reporting-lua
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply README.md:1]: Fixed.

File: README.md (hunk 2 above)
Comment (redcatbear) [reply README.md:22]: Fixed.

File: doc/user_guide/user_guide.md (hunk 8 above)
Comment (redcatbear) [reply doc/user_guide/user_guide.md:13]: Change explanation term to "identifier".

File: doc/changes/changelog.md (hunk 1 above)
Comment (redcatbear) [reply doc/changes/changelog.md:1]: Fixed.

File: README.md (hunk 3 above)
Comment (redcatbear) [reply README.md:28]: Fixed.

File: README.md (hunk 4 above)
Comment (redcatbear) [reply README.md:41]: Fixed.

File: .github/workflows/build.yml [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    steps:
+      - name: "Install dependencies"
+        run: |
+          sudo apt-get update
+          sudo apt install luarocks
+          sudo luarocks install luacheck
+          sudo luarocks install luacov
+          sudo luarocks install luacov-coveralls
+          sudo luarocks install LuaUnit
+          sudo luarocks install Mockagne
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply .github/workflows/build.yml]: I use a combination now. LuaRocks 2 only support runtime dependency definition, so the build dependencies still need to be installed extra. Will switch once LuaRocks 3 is the default on Ubuntu.

File: README.md (hunk 5 above)
Comment (redcatbear) [reply README.md]: Removed.

File: doc/changes/changes_1.0.0.md (hunk 7 above)
Comment (redcatbear) [reply doc/changes/changes_1.0.0.md:3]: Looks like I taught you well. Was curious if you would find this. :-)

File: doc/user_guide/user_guide.md (hunk 9 above)
Comment (redcatbear) [reply doc/user_guide/user_guide.md:54]: Create #3.

File: doc/user_guide/user_guide.md (hunk 10 above)
Comment (redcatbear) [reply doc/user_guide/user_guide.md:53]: Created #4.

File: doc/user_guide/user_guide.md (hunk 11 above)
Comment (redcatbear) [reply doc/user_guide/user_guide.md:99]: I forgot to mention another option that allows this.

Comment (redcatbear) [reply doc/user_guide/user_guide.md:99]: Added to documentation.

File: src/exaerror.lua (hunk 15 above)
Comment (redcatbear) [reply src/exaerror.lua:47]: Because otherwise you get an error that you tried to append a raw table (the error object). This way it works just fine.

File: target/luaunit-reports/test_errormsg.xml (hunk 16 above)
Comment (redcatbear) [reply target/luaunit-reports/test_errormsg.xml:1]: Added to .gitignore (was autogenerated by GitHub).

File: src/exaerror.lua (hunk 14 above)
Comment (redcatbear) [reply src/exaerror.lua:29]: I think the edge case is fine here. The message about a missing message leaves to much irritation on the end users part in my eyes.

File: src/exaerror.lua (hunk 13 above)
Comment (redcatbear) [reply src/exaerror.lua:24]: Hm. I admit that this is more readable. At the expense of containing newline characters which are always a problem in logs.

Comment (redcatbear) [reply src/exaerror.lua:24]: Changed.

File: src/exaerror.lua (hunk 12 above)
Comment (redcatbear) [reply src/exaerror.lua:23]: Iteration with `in` requires an iterator function.

For mitigation parameter replacement I created #3.

File: target/luaunit-reports/test_errormsg.xml (hunk 16 above)
Comment (jakobbraun) [reply target/luaunit-reports/test_errormsg.xml:1]: But it's still there....

Comment (jakobbraun) [reply target/luaunit-reports/test_errormsg.xml:1]: I think you have to remove it if it was added to git once

File: target/luaunit-reports/test_exaerror.xml [hunk 19]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (jakobbraun) [inline target/luaunit-reports/test_exaerror.xml:1]: same

File: test/test_exaerror.lua [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+        function ()
+            error(exaerror.create("E-IO-13", "Need %.1f MiB space, but only %.1f MiB left on device %s.",
+                500.2, 14.8, "/dev/sda4"))
+        end
+    )
+end
+
+function test_exaerror.test_throw_error_directly()
+    luaunit.assertErrorMsgContains(
+        "E-IO-13: Need 500.2 MiB space, but only 14.8 MiB left on device /dev/sda4.",
```
Comment (jakobbraun) [inline test/test_exaerror.lua:109]: Nice examples

File: tools/runtests [hunk 21]
```diff
@@ -0,0 +1,114 @@
+#!/bin/bash
```
Comment (jakobbraun) [inline tools/runtests:1]: Maybe add shell-check to the CI

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: tools/runtests (hunk 21 above)
Comment (redcatbear) [reply tools/runtests:1]: Good idea. Done.

File: .project [hunk 22]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply .project]: Removed.

**Review by jakobbraun — APPROVED** [review]
